### PR TITLE
expression,util/stringutil: move LIKE pattern match to stringutil package

### PR DESCRIPTION
**Pull request by tiancaiamao**
This code will be reused, so it's better put it to stringutil.

### CODE REVIEW

**Comment by hanfei1991** [comment]
I think you can add some tests?

**Comment by tiancaiamao** [comment]
Good idea!     I'll do it.

**Comment by tiancaiamao** [comment]
Updated, PTAL @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

File: util/stringutil/string_util.go [hunk 1]
```diff
@@ -131,3 +131,111 @@ func Unquote(s string) (t string, err error) {
...
+			if i < len(pattern)-1 {
+				i++
+				c = pattern[i]
+				if c == escape || c == '_' || c == '%' {
+					// valid escape.
+				} else {
+					// invalid escape, fall back to escape byte
+					// mysql will treat escape character as the origin value even
+					// the escape sequence is invalid in Go or C.
+					// e.g, \m is invalid in Go, but in MySQL we will get "m" for select '\m'.
```
Comment (zimulala) [inline util/stringutil/string_util.go]: s/e.g,/e.g.

File: util/stringutil/string_util.go [hunk 2]
```diff
@@ -131,3 +131,111 @@ func Unquote(s string) (t string, err error) {
...
+		var c = pattern[i]
+		switch c {
+		case escape:
+			lastAny = false
+			tp = patMatch
+			if i < len(pattern)-1 {
+				i++
+				c = pattern[i]
+				if c == escape || c == '_' || c == '%' {
+					// valid escape.
```
Comment (zimulala) [inline util/stringutil/string_util.go]: s/valid/Valid

File: util/stringutil/string_util.go [hunk 3]
```diff
@@ -131,3 +131,111 @@ func Unquote(s string) (t string, err error) {
...
+				} else {
+					// invalid escape, fall back to escape byte
+					// mysql will treat escape character as the origin value even
+					// the escape sequence is invalid in Go or C.
+					// e.g, \m is invalid in Go, but in MySQL we will get "m" for select '\m'.
+					// Following case is correct just for escape \, not for others like +.
+					// TODO: add more checks for other escapes.
```
Comment (zimulala) [inline util/stringutil/string_util.go]: s/add/Add

File: util/stringutil/string_util.go [hunk 4]
```diff
@@ -131,3 +131,111 @@ func Unquote(s string) (t string, err error) {
...
+		case escape:
+			lastAny = false
+			tp = patMatch
+			if i < len(pattern)-1 {
+				i++
+				c = pattern[i]
+				if c == escape || c == '_' || c == '%' {
+					// valid escape.
+				} else {
+					// invalid escape, fall back to escape byte
```
Comment (zimulala) [inline util/stringutil/string_util.go]: s/invalid/Invalid

**Comment by zimulala** [comment]
Rest LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**